bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieProtoEncoding.java

```diff
@@ -289,8 +289,8 @@ public Object decode(ByteBuf buffer)
                 entryId = buffer.readLong();
 
                 if (rc == BookieProtocol.EOK) {
-                    ByteBuf content = buffer.slice();
-                    return new BookieProtocol.ReadResponse(version, rc, ledgerId, entryId, content.retain());
+                    return new BookieProtocol.ReadResponse(version, rc,
```

Comment:
I'm not 100% sure the `slice()` can be avoided, because we would have to double check all the places that are touching that buffer.

I think a better option here is to use `buffer.retainedSlice()` which achieves both goals (and also lets the resulting `ByteBuf` slice object to be pooled.

Reply:
`retainedSlices()` was not used originally because it was introduced in netty-4.1 while the original patch was based on netty-4.0.

Reply:
Alright, I'll switch it to retainedSlice()